### ISSUE TITLE
The allowed os/arch combinations are too restrictive

### ISSUE THREAD

**Issue by jonjohnsonjr** [estesp/manifest-tool#108]
I stumbled across this hardcoded list in manifest-tool while reviewing something unrelated: https://github.com/estesp/manifest-tool/blob/2d360eeba276afaf63ab22270b7c6b4f8447e261/pkg/util/os.go#L5-L41

[Per the spec](https://github.com/opencontainers/image-spec/blob/master/image-index.md#image-index-property-descriptions), `os` and `architecture` values come from GOOS/GOARCH, while the `variant` field is hardcoded.

From what I understand, it doesn't make sense to validate the combinations like this (other than for arm + variant). This list of valid combinations are specific to the go compiler and completely unrelated to the image spec. As is, manifest-tool is only capable of pushing images for platforms for which the golang compiler has an implementation. I think (if you're going to be doing this validation at all) you should just be checking that:
1. `architecture` is a valid GOARCH, and
2. `os` is a valid GOOS, and 
3. `variant` is valid per the spec:
    1. `v6` == `arm`
    2. `v7` == `arm`
    2. `v8` == `arm`|`arm64`

**Comment by estesp** [comment]
Good catch; I've opened #110 to solve this, at least in the sense that I can make it part of release to run `go generate` to get the current supported list from Go. If there is a better way to handle this at runtime I'm open to ideas.

**Comment by jonjohnsonjr** [comment]
> If there is a better way to handle this at runtime I'm open to ideas.

So the main issue isn't that this was hardcoded, it's that you're checking the GOOS/GOARCH combination intead of just GOOS and GOARCH separately.

Instead of this:
```go
_, ok := validOSArch[osarch]
```

You want something like:
```go
_, okOS := validOS[os]
_, okArch := validArch[arch]

// then the arm stuff

return okOS && okArch
```

These values being valid is unrelated to whether or not golang can compile that particular combination (which is what you're checking atm).

**Comment by estesp** [comment]
@jonjohnsonjr thanks for the clarification. I think what I was missing is the difference between "spec compliant" and trying to do something along the lines of input validation.

I'm leaning towards this particular tool being more valuable with some validation for the end user that what they are submitting as a platform object has "reasonableness" from the this-possibly-could-work standpoint. The fact that "windows/s390x" might be valid from a pure spec compliance standpoint misses the point that someone assembling yaml might be copying sections and forget to change the OS as they increment through a set of CPU architectures, and "valid per the spec" in that case is unhelpful versus the opportunity to catch the fact that they now are going to write a multi-platform index that won't ever trigger for the platforms they expected.

If you think there is a use case where the current list (which is a restriction from "all possible spec compliant values" for sure) actually limits the ability of someone to create a working manifest list/index I'm definitely open to understanding. I realize a more generic tool without the express purpose of creating working manifest list/index images might lean towards spec validity over "catching user error."

**Comment by jonjohnsonjr** [comment]
tl;dr there are valid platforms that go cannot compile and invalid platforms that go can compile, so it's not a great proxy for validating a container platform

> some validation for the end user that what they are submitting as a platform object has "reasonableness" from the this-possibly-could-work standpoint

From this perspective, what you have is not restrictive enough 😄 go can compile for a lot of platforms that probably won't be running a container.

I've tried to explain with a terrible diagram:

![image](https://user-images.githubusercontent.com/17863526/107288187-e5f89e00-6a17-11eb-8c8c-e662953839ab.png)

The black outer circle is what the spec permits, and what I'm proposing.

The red is what go can compile (and what `manifest-tool` accepts as valid).

The blue circle is hard to define, but there is definitely some set of "likely" container platforms.

The `manifest-tool` validation will gladly let me set the following platforms that (IMO) don't make much sense:

```
android/386
android/amd64
android/arm
android/arm64
darwin/amd64
darwin/arm64
js/wasm
```

(Go 1.16 adds `ios/arm64` as well, since `darwin/arm64` will be taken by M1.)

_Maybe_ android makes sense, but I don't think darwin or ios do? Not to mention plan 9, AIX, solaris, illumos, and all the BSD operating systems for which go has compiler support.

The go compiler dropped support for `darwin/386` after 1.14 (https://github.com/golang/go/issues/34749) which means this tool would drop support for producing manifest lists that point to `darwin/386`. If you feel like that's fine because nobody is producing `darwin` containers,  I think it's a bit strange that `darwin/amd64` _does_ validate successfully... you know?

> If you think there is a use case where the current list (which is a restriction from "all possible spec compliant values" for sure) actually limits the ability of someone to create a working manifest list/index I'm definitely open to understanding.

Theoretically, if I wanted to run a container on, say, [`windows/arm64`](https://github.com/microsoft/openjdk-aarch64), I am currently unable to use `manifest-tool` to add it to a manifest list (at least as of go 1.15.5):

```
$ go tool dist list | grep windows
windows/386
windows/amd64
windows/arm
```

**Comment by estesp** [comment]
Yup, totally understand. In a simpler time the Go-as-a-proxy for container platforms was a pretty reasonable bet, or at least the Venn diagram had enough overlap that it was "good enough".

My general data-agnostic 😇 hope was that the blue "outside the red" overlap was incredibly niche for now, but maybe `windows/arm64` is a good enough example that it isn't. And you're right, validation against all-possible-Go-compilable platforms is getting pretty far afield of "reasonable container platforms" (although counterpoint for darwin/amd64: it's actually getting close to some definition of reasonableness; a FOSDEM talk this weekend showed the status of the darwin port, not just to build containerd but for running "containers" on the platform). 😂 

Anyway, I think I'm going to agree that the validation of OS and arch as separate "validity checks" gets us to a useful compromise for now. I'll update the PR

**Comment by jonjohnsonjr** [comment]
> a FOSDEM talk this weekend showed the status of the darwin port, not just to build containerd but for running "containers" on the platform

Ooh interesting... [this one](https://fosdem.org/2021/schedule/event/containers_darwin_containerd/)? This seems to just run linux containers on darwin, from the abstract, but I haven't watched the talk yet.

> Anyway, I think I'm going to agree that the validation of OS and arch as separate "validity checks" gets us to a useful compromise for now. I'll update the PR

🎉 

I definitely understand the logic from before, but this feels like the better of two imperfect options :P
